libraries/chain/include/graphene/chain/hardfork_visitor.hpp

```diff
@@ -0,0 +1,83 @@
+#pragma once
+/*
+ * Copyright (c) 2019 Contributors
+ *
+ * The MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <graphene/protocol/operations.hpp>
+
+#include <graphene/chain/hardfork.hpp>
+
+#include <fc/reflect/typelist.hpp>
+
+#include <type_traits>
+#include <functional>
+
+namespace graphene { namespace chain {
+using namespace protocol;
+namespace TL { using namespace fc::typelist; }
+
+/**
+ * @brief The hardfork_visitor struct checks whether a given operation type has been hardforked in or not
+ *
+ * This visitor can be invoked in several different ways, including operation::visit, typelist::runtime::dispatch, or
+ * direct invocation by calling the visit() method passing an operation variant, narrow operation type, operation tag,
+ * or templating on the narrow operation type
+ */
+struct hardfork_visitor {
+   using result_type = bool;
+   using first_unforked_op = custom_authority_create_operation;
+   using BSIP_40_ops = TL::list<custom_authority_create_operation, custom_authority_update_operation,
+                                custom_authority_delete_operation>;
+   fc::time_point_sec now;
+
+   hardfork_visitor(fc::time_point_sec now) : now(now) {}
+
+   /// The real visitor implementations. Future operation types get added in here.
+   /// @{
+   template<typename Op>
+   std::enable_if_t<operation::tag<Op>::value < operation::tag<first_unforked_op>::value, bool>
+   visit() { return true; }
+   template<typename Op>
+   std::enable_if_t<TL::contains<BSIP_40_ops, Op>(), bool>
+   visit() { return HARDFORK_BSIP_40_PASSED(now); }
```

Comment:
I think it's better to put the code related to bsip40 in a new bsip40-hf-visitor class/struct which is derived from a generic hf-visitor class/struct.

Reply:
The `hardfork_visitor` is quite similar to the `get_impacted_account_visitor` -- it should work for all `operation` types and should tell you something about that operation in a reasonably polymorphic fashion. Thus it needs to deal with all operation types within a single struct; it can't reasonably be split over multiple structs or files.

It looks like the BSIP40 code is a special case here because it's the only case that existed at the time it was written that wasn't trivially true. Over time, all new operations will be added to this so that a generic `operation` can always be inspected to see if it's forked in yet or not, so the BSIP40 handling is not really out of place here.